src/core/context-management/ContextManager.ts

```diff
@@ -9,57 +9,76 @@ import * as path from "path"
 import fs from "fs/promises"
 import cloneDeep from "clone-deep"
 
-// array of string values allows us to cover all changes currently
-export type MessageContent = string[]
+enum EditType {
+	UNDEFINED = 0,
+	NO_FILE_READ = 1,
+	READ_FILE_TOOL = 2,
+	ALTER_FILE_TOOL = 3,
+	FILE_MENTION = 4,
+}
+
+// array of string values allows us to cover all changes for message types currently supported
+type MessageContent = string[]
+type MessageMetadata = string[][]
 
 // Type for a single context update
-type ContextUpdate = [number, string, MessageContent] // [timestamp, updateType, update]
+type ContextUpdate = [number, string, MessageContent, MessageMetadata] // [timestamp, updateType, update, metadata]
 
 // Type for the serialized format of our nested maps
 type SerializedContextHistory = Array<
 	[
 		number, // messageIndex
-		Array<
-			[
-				number, // blockIndex
-				ContextUpdate[], // updates array
-			]
-		>,
+		[
+			number, // EditType (message type)
+			Array<
+				[
+					number, // blockIndex
+					ContextUpdate[], // updates array (now with 4 elements including metadata)
+				]
+			>,
+		],
 	]
 >
 
 export class ContextManager {
 	// mapping from the apiMessages outer index to the inner message index to a list of actual changes, ordered by timestamp
 	// timestamp is required in order to support full checkpointing, where the changes we apply need to be able to be undone when
 	// moving to an earlier conversation history checkpoint - this ordering intuitively allows for binary search on truncation
+	// there is also a number stored for each (EditType) which defines which message type it is, for custom handling
 
-	// format:  {outerIndex => {innerIndex => [[timestamp, updateType, update], ...]}}
-	// example: { 1 => { 0 => [[<timestamp>, "text", "[NOTE] Some previous conversation history with the user has been removed ..."], ...] } }
+	// format:  { outerIndex => [EditType, { innerIndex => [[timestamp, updateType, update], ...] }] }
+	// example: { 1 => { [0, 0 => [[<timestamp>, "text", "[NOTE] Some previous conversation history with the user has been removed ..."], ...] }] }
 	// the above example would be how we update the first assistant message to indicate we truncated text
-	private contextHistoryUpdates: Map<number, Map<number, ContextUpdate[]>>
+	private contextHistoryUpdates: Map<number, [number, Map<number, ContextUpdate[]>]>
```

Comment:
This should be typed to the ENUM or a similar struct that allows this to be understood from the code alone.  I do not now what "number" represents here from the typing.